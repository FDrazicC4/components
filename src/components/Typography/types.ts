import { ReactNode } from 'react';


export type Variants = 'h1' | 'h2' | 'h3' | 'body1' | 'body2' | 'label' | 'inherit';
export type Colors = (
    'primary'
    | 'secondary'
    | 'tertiary'
    | 'success'
    | 'danger'
    | 'warning'
    | 'info'
    | 'light'
    | 'dark'
    | 'inherit'
);
export type Alignments = 'inherit' | 'left' | 'center' | 'right' | 'justify';
export type Displays = 'initial' | 'block' | 'inline' | 'inline-block';


export interface Props {

    /**
     * The content or text to be rendered. Usually just a string, but can be any valid JSX.
     */
    children: ReactNode,

    /**
     * Classes to be passed through to the component.
     */
    className?: string,

    /**
     * The overall style to be applied to the typography.
     */
    variant?: Variants,

    /**
     * The font color of the content.
     */
    color?: Colors,

    /**
     * The text alignment of the content.
     */
    align?: Alignments,

    /**
     * Can be used to override the display type of the wrapping element.
     */
    display?: Displays,

    /**
     * Adds text truncation (ellipsis) to the wrapping element.
     */
    shouldTruncate?: boolean,

    /**
     * Can be used to override the element used to wrap the content.
     */
    as?: keyof JSX.IntrinsicElements | null,

    /**
     * A tab index to be passed to the underlying element
     */
    tabIndex?: number,

}
