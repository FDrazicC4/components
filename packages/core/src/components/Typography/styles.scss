@import url('https://fonts.googleapis.com/css?family=Roboto:300,400,500,700');
@import "../../styles/utilities";


// ==============================================
// General
// ==============================================

.wrapper {
    font-family: Roboto, serif;
    margin: 0;
}


// ==============================================
// Colors
// ==============================================

$colors: primary, secondary, tertiary, success, info, warning, danger;

@mixin color-styles($color) {
    color: map_get(map_get($typography, colors), $color);
}

@each $color in $colors {
    .wrapper.color--#{$color} {
        @include color-styles($color);
    }
}


// ==============================================
// Variants
// ==============================================

$variants: h1, h2, h3, body1, body2, label;

@mixin variant-styles($variant) {
    font-size: map_get(map_get($typography, font-sizes), $variant);
    font-weight: map_get(map_get($typography, font-weights), $variant);
    line-height: map_get(map_get($typography, line-heights), $variant);
    color: map_get(map_get($typography, colors), $variant);
}

@each $variant in $variants {
    .variant--#{$variant} {
        @include variant-styles($variant);
    }
}

.variant--label {
    text-transform: uppercase;
}

.variant--inherit {
    font-size: inherit;
}


// ==============================================
// Alignments
// ==============================================

$alignments: inherit, left, right, center, justify;

@each $alignment in $alignments {
    .align--#{$alignment} {
        text-align: $alignment;
    }
}


// ==============================================
// Display types
// ==============================================

$displays: block, inline, inline-block;

@each $display in $displays {
    .display--#{$display} {
        display: $display;
    }
}


// ==============================================
// States
// ==============================================

.hasBottomMargin {
    margin-bottom: 1rem;
}

.shouldTruncate {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
}
