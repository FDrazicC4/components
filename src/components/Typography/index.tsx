import React from 'react';
import clsx from 'clsx';
import { Typography as MuiTypography } from '@material-ui/core';
import { Props } from './types';



/**
 * The `<Typography />` component renders text using the 'Roboto' font in various styles.
 */
export const Typography = React.forwardRef(({
    variant = 'body1',
    align = 'inherit',
    display = 'initial',
    shouldTruncate = false,
    color,
    hasGutter = false,
    className,
    ...otherProps
}: Props, ref: React.Ref<any>): React.ReactElement => {


    const classNames = clsx(
        'brew-Typography',
        [`brew-Typography--display-${display}`],
        { [`brew-Typography--variant-${String(variant)}`]: variant },
        { [`brew-Typography--align-${String(align)}`]: align },
        { [`brew-Typography--color-${String(color)}`]: color },
        { 'brew-Typography--shouldTruncate': shouldTruncate },
        { 'brew-Typography--hasGutter': hasGutter },
        className,
    );


    return (
        <MuiTypography
            align={align}
            className={classNames}
            display={display}
            gutterBottom={hasGutter}
            noWrap={shouldTruncate}
            ref={ref}
            variant={variant}
            {...otherProps}
        />
    );


});


Typography.displayName = 'Typography';


export default Typography;
