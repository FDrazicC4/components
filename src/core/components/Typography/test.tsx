import React from 'react';
import { shallow } from 'enzyme';
import Type from './index';


it('renders children when passed in', () => {
    const wrapper = shallow((
        <Type>
            <span>foo</span>
        </Type>
    ));
    expect(wrapper.contains(<span>foo</span>)).toEqual(true);
});


it('accepts custom classes', () => {
    const wrapper = shallow((
        <Type className="customClass" />
    ));
    expect(wrapper.hasClass('customClass')).toEqual(true);
    expect(wrapper.hasClass('wrapper')).toEqual(true);
});


it('"variant" defaults to "inherit" and applies appropriate class', () => {
    const wrapper = shallow((
        <Type />
    ));
    expect(wrapper.hasClass('variant--inherit')).toEqual(true);
});


it('"color" defaults to "initial" and applies appropriate class', () => {
    const wrapper = shallow((
        <Type />
    ));
    expect(wrapper.hasClass('color--initial')).toEqual(true);
});


it('"display" defaults to "initial" and applies appropriate class', () => {
    const wrapper = shallow((
        <Type />
    ));
    expect(wrapper.hasClass('display--initial')).toEqual(true);
});


it('"align" defaults to "inherit" and applies appropriate class', () => {
    const wrapper = shallow((
        <Type />
    ));
    expect(wrapper.hasClass('align--inherit')).toEqual(true);
});