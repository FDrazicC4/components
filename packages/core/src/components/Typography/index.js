import React from 'react';
import clsx from 'clsx';
import { Props } from './types';
import styles from './styles.scss';


function Type(props: Props) {


    const {
        variant = 'inherit',
        color = 'initial',
        display = 'initial',
        align = 'inherit',
        hasBottomMargin,
        shouldTruncate,
        className,
        ...otherProps
    } = props;


    const classes = clsx(
        styles.wrapper,
        styles[`variant--${variant}`],
        styles[`color--${color}`],
        styles[`display--${display}`],
        styles[`align--${align}`],
        { [styles.hasBottomMargin]: hasBottomMargin },
        { [styles.shouldTruncate]: shouldTruncate },
        className,
    );


    let TextElement = 'span';
    if (variant === 'h1' || variant === 'h2' || variant === 'h3') TextElement = variant;


    return (
        <TextElement className={classes} {...otherProps} />
    );


}


export default Type;
