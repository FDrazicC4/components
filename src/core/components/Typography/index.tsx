import React from 'react';
import clsx from 'clsx';
import { Props } from './types';
import styles from './styles.scss';


/**
 * The `Type` component renders text using the 'Roboto' font in various styles.
 */
function Typography({
    children = null,
    variant = 'inherit',
    color = 'initial',
    align = 'inherit',
    display = 'initial',
    hasBottomMargin = false,
    shouldTruncate = false,
    className = null,
    as = null,
    ...otherProps
}: Props) {


    const classes = clsx(
        styles.wrapper,
        styles[`variant--${variant}`],
        styles[`color--${color}`],
        styles[`align--${align}`],
        styles[`display--${display}`],
        { [styles.hasBottomMargin]: hasBottomMargin },
        { [styles.shouldTruncate]: shouldTruncate },
        className,
    );


    let TextElement: keyof JSX.IntrinsicElements = 'span';
    if (as) TextElement = as;
    else if (variant === 'h1' || variant === 'h2' || variant === 'h3') TextElement = variant as 'div';


    return (
        <TextElement className={classes} {...otherProps}>
            {children}
        </TextElement>
    );


}


export default Typography;
