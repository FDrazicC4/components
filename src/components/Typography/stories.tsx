import React, { ReactElement, ReactNode } from 'react';
import { text, select, boolean } from '@storybook/addon-knobs';
import Label from '../Label';
import Typography from './index';
import { Variants, Colors, Alignments, Displays } from './types';


export default {
    component: Typography,
    parameters: {
        componentSubtitle: <Label color="success">Stable</Label>,
    },
    title: 'Components|Display/Typography',
};


export const General = (): ReactElement => {
    const align: Alignments = select('align', ['inherit', 'left', 'center', 'right', 'justify'], 'inherit');
    const children: ReactNode = text('children', 'Lorem ipsum dolor sit amet.');

    const color: Colors = select('color', [
        'initial',
        'primary',
        'secondary',
        'tertiary',
        'success',
        'warning',
        'info',
        'danger',
    ], 'initial');
    const display: Displays = select('display', ['initial', 'block', 'inline', 'inline-block'], 'initial');
    const shouldTruncate = boolean('shouldTruncate', false);
    const variant: Variants = select('variant', ['h1', 'h2', 'h3', 'body1', 'body2', 'label', 'inherit'], 'h1');

    return (
        <Typography
            align={align}
            color={color}
            display={display}
            shouldTruncate={shouldTruncate}
            variant={variant}
        >
            {children}
        </Typography>
    );
};


export const Default = (): ReactElement => (
    <div style={{ maxWidth: '100%' }}>
        <Typography>
            Default Settings - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </Typography>
    </div>
);

export const Headers = (): ReactElement => (
    <div style={{ maxWidth: '100%' }}>
        <Typography shouldTruncate variant="h1">
                H1 - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </Typography>
        <Typography shouldTruncate variant="h2">
                H2 - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </Typography>
        <Typography shouldTruncate variant="h3">
                H3 - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </Typography>
    </div>
);


export const Paragraphs = (): ReactElement => (
    <React.Fragment>
        <Typography as="p" variant="body1">
            body1 - Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed eu ante hendrerit, imperdiet lectus in,
            egestas nulla. Donec molestie congue nulla sit amet viverra. Duis dignissim mi eget ex dignissim, id
            feugiat risus dapibus. Ut auctor imperdiet turpis, vel mollis dolor sollicitudin in.
        </Typography>
        <Typography as="p" variant="body2">
            body1 - Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed eu ante hendrerit, imperdiet lectus in,
            egestas nulla. Donec molestie congue nulla sit amet viverra. Duis dignissim mi eget ex dignissim, id
            feugiat risus dapibus. Ut auctor imperdiet turpis, vel mollis dolor sollicitudin in.
        </Typography>
    </React.Fragment>
);


export const Truncated = (): ReactElement => (
    <Typography as="p" shouldTruncate variant="body1">
            body1 - Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed eu ante hendrerit, imperdiet lectus in,
            egestas nulla. Donec molestie congue nulla sit amet viverra. Duis dignissim mi eget ex dignissim, id
            feugiat risus dapibus. Ut auctor imperdiet turpis, vel mollis dolor sollicitudin in.
    </Typography>
);


export const Labels = (): ReactElement => (
    <Typography as="p" shouldTruncate variant="label">
            label - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
    </Typography>
);

