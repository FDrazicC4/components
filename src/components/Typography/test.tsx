import React from 'react';
import renderer from 'react-test-renderer';
import Typography from './index';


it('renders correctly', () => {
    const tree = renderer
        .create(<Typography>Foo</Typography>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});


it('renders correctly with various props', () => {
    const tree = renderer
        .create((
            <React.Fragment>
                <Typography align="left" className="foo" color="primary" shouldTruncate>foo</Typography>
                <Typography align="right" color="secondary">foo</Typography>
                <Typography align="center" color="tertiary" variant="h4">foo</Typography>
            </React.Fragment>
        ))
        .toJSON();
    expect(tree).toMatchSnapshot();
});
