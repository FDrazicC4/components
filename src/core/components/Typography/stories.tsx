import React, { ReactNode } from 'react';
import { text, select, boolean } from '@storybook/addon-knobs';
import Label from '../Label';
import Typography from './index';
import { Variants, Colors, Alignments, Displays } from './types';


export default {
    title: 'Components|Display/Typography',
    component: Typography,
    parameters: {
        componentSubtitle: <Label color="success">Stable</Label>,
    },
};


export const General = () => {
    const align: Alignments = select('align', ['inherit', 'left', 'center', 'right', 'justify'], 'inherit');
    const children: ReactNode = text('children', 'Lorem ipsum dolor sit amet.');
    const color: Colors = select('color', ['initial', 'primary', 'secondary', 'tertiary', 'success', 'warning', 'info', 'danger'], 'initial');
    const display: Displays = select('display', ['initial', 'block', 'inline', 'inline-block'], 'initial');
    const hasBottomMargin = boolean('hasBottomMargin', true);
    const shouldTruncate = boolean('shouldTruncate', false);
    const variant: Variants = select('variant', ['h1', 'h2', 'h3', 'body1', 'body2', 'label', 'inherit'], 'h1');

    return (
        <React.Fragment>
            <Typography
                align={align}
                color={color}
                display={display}
                hasBottomMargin={hasBottomMargin}
                shouldTruncate={shouldTruncate}
                variant={variant}
            >
                {children}
            </Typography>
        </React.Fragment>
    );
};


export const Headers = () => (
    <React.Fragment>
        <div style={{ maxWidth: '100%' }}>
            <Typography variant="h1" shouldTruncate hasBottomMargin>
                H1 - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
            </Typography>
            <Typography variant="h2" shouldTruncate hasBottomMargin>
                H2 - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
            </Typography>
            <Typography variant="h3" shouldTruncate hasBottomMargin>
                H3 - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
            </Typography>
        </div>
    </React.Fragment>
);


export const Paragraphs = () => (
    <React.Fragment>
        <Typography variant="body1" as="p" hasBottomMargin>
            body1 - Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed eu ante hendrerit, imperdiet lectus in,
            egestas nulla. Donec molestie congue nulla sit amet viverra. Duis dignissim mi eget ex dignissim, id
            feugiat risus dapibus. Ut auctor imperdiet turpis, vel mollis dolor sollicitudin in.
        </Typography>
        <Typography variant="body2" as="p" hasBottomMargin>
            body1 - Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed eu ante hendrerit, imperdiet lectus in,
            egestas nulla. Donec molestie congue nulla sit amet viverra. Duis dignissim mi eget ex dignissim, id
            feugiat risus dapibus. Ut auctor imperdiet turpis, vel mollis dolor sollicitudin in.
        </Typography>
    </React.Fragment>
);


export const Truncated = () => (
    <React.Fragment>
        <Typography variant="body1" as="p" shouldTruncate>
            body1 - Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed eu ante hendrerit, imperdiet lectus in,
            egestas nulla. Donec molestie congue nulla sit amet viverra. Duis dignissim mi eget ex dignissim, id
            feugiat risus dapibus. Ut auctor imperdiet turpis, vel mollis dolor sollicitudin in.
        </Typography>
    </React.Fragment>
);


export const Labels = () => (
    <React.Fragment>
        <Typography variant="label" as="p" shouldTruncate hasBottomMargin>
            label - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </Typography>
    </React.Fragment>
);

