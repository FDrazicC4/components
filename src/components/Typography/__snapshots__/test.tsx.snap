// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<span
  className="wrapper variant--inherit color--initial align--inherit display--initial"
>
  Foo
</span>
`;
